kernel/core/datamanagement/memory.cpp

```diff
@@ -3820,6 +3820,226 @@ std::vector<AnovaResult> Memory::getAnova(const VectorIndex& colCategories, size
 }
 
 
+/////////////////////////////////////////////////
+/// \brief check if value type vector is already in a vector of vectors
+///
+/// \param newVector const std::vector<mu::value_type>&
+/// \param centroids const std::vector<std::vector<mu::value_type>>&
+/// \return bool
+///
+/////////////////////////////////////////////////
+static bool isUnique(const std::vector<mu::value_type>& newVector, const std::vector<std::vector<mu::value_type>>& centroids)
+{
+    return std::none_of(centroids.begin(), centroids.end(),
+                        [&newVector](const std::vector<mu::value_type>& centroid)
+                        { return newVector == centroid; });
+}
+
+
+/////////////////////////////////////////////////
+/// \brief Calculate L2 Norm of 2 given value type vectors
+///
+/// \param vec1 const std::vector<mu::value_type>&
+/// \param vec2 const std::vector<mu::value_type>&
+/// \return double
+///
+/////////////////////////////////////////////////
+static double calculateL2Norm(const std::vector<mu::value_type>& vec1, const std::vector<mu::value_type>& vec2)
+{
+
+    mu::value_type sum = 0.0;
+    for (size_t i = 0; i < vec1.size(); ++i)
+    {
+        sum += (vec1[i] - vec2[i]) * (vec1[i] - vec2[i]);
+    }
+
+    // todo do we have sqrt f�r value_type
+    return std::sqrt(sum.real());
+}
+
+
+/////////////////////////////////////////////////
+/// \brief get all Indices of elements with given value
+///
+/// \param vec const std::vector<mu::value_type>&
+/// \param value mu::value_type
+/// \return std::vector<int>
+///
+/////////////////////////////////////////////////
+static std::vector<int> getIndices(const std::vector<mu::value_type>& vec, mu::value_type value)
+{
+    std::vector<int> indices;
+    auto it = vec.begin();
+
+    while ((it = std::find(it, vec.end(), value)) != vec.end())
+    {
+        indices.push_back(std::distance(vec.begin(), it));
+        ++it;  // Move past the last found element
+    }
+
+    return indices;
+}
+
+/////////////////////////////////////////////////
+/// \brief calculate kmeans
+///
+/// \param columns const VectorIndex&
+/// \param nClusters size_t
+/// \param maxIterations size_t
+/// \param init_method size_t    if 2 kmeans++ is used, random seeds otherwise
+/// \return KMeansResult
+///
+/////////////////////////////////////////////////
+KMeansResult Memory::getKMeans(const VectorIndex& columns, size_t nClusters, size_t maxIterations, size_t init_method) const
+{
+    std::vector<mu::value_type> clusters;
+
+    // Ensure that we have data
+    if (memArray.size() <= columns[0] || !memArray[columns[0]])
+    {
+        clusters.resize(memArray[columns[0]]->size(), NAN);
+        return KMeansResult();
+    }
+
+    //Memory _mem(columns.size());
+    // todo is there a more efficient way utilizing the Memory class?
+    // -> currently always get element as vector from Memory class
+    // -> is there a better way to access each data row ?

```

Comment:
Well, `readMem()` as you already found, didn't you? The data model is built columnwise for good reasons, therefore actions on columns are fast, but rows are harder and slower.